source/s3_auto_ranged_put.c

```diff
@@ -1184,72 +1154,52 @@ static void s_s3_auto_ranged_put_request_finished(
     }
 }
 
-static int s_s3_auto_ranged_put_pause(struct aws_s3_meta_request *meta_request, struct aws_string **out_resume_token) {
+static int s_s3_auto_ranged_put_pause(
+    struct aws_s3_meta_request *meta_request,
+    struct aws_s3_meta_request_resume_token **out_resume_token) {
+
+    int return_status = AWS_OP_SUCCESS;
+    *out_resume_token = NULL;
 
     /* lock */
     aws_s3_meta_request_lock_synced_data(meta_request);
     struct aws_s3_auto_ranged_put *auto_ranged_put = meta_request->impl;
 
-    /* only generate token if multipart upload was completed, since there is nothing to resume otherwise*/
-    int token_generated_error = AWS_OP_SUCCESS;
-    *out_resume_token = NULL;
-    if (auto_ranged_put->synced_data.create_multipart_upload_completed) {
-        struct aws_json_value *root = aws_json_value_new_object(meta_request->allocator);
-
-        /**
-         * generate pause token for put.
-         * current format is json with the following structure
-         * {
-         *      type: String
-         *      multipart_upload_id: String
-         *      partition_size: number
-         *      total_num_parts: number
-         * }
-         */
-
-        aws_json_value_add_to_object(
-            root,
-            aws_byte_cursor_from_c_str("type"),
-            aws_json_value_new_string(
-                meta_request->allocator, aws_byte_cursor_from_c_str("AWS_S3_META_REQUEST_TYPE_PUT_OBJECT")));
-
-        aws_json_value_add_to_object(
-            root,
-            aws_byte_cursor_from_c_str("multipart_upload_id"),
-            aws_json_value_new_string(
-                meta_request->allocator, aws_byte_cursor_from_string(auto_ranged_put->upload_id)));
-
-        aws_json_value_add_to_object(
-            root,
-            aws_byte_cursor_from_c_str("partition_size"),
-            aws_json_value_new_number(meta_request->allocator, (double)meta_request->part_size));
+    /* upload can be in one of several states:
+     *   - not started, i.e. we didn't even call crete mpu yet - return success,
+     *     token is NULL and cancel the upload
+     *   - in the middle of upload - return success, create token and cancel
+     *     upload
+     *   - after upload - return already completed error, token is NULL and
+     *     nothing to cancel
+     */
 
-        aws_json_value_add_to_object(
-            root,
-            aws_byte_cursor_from_c_str("total_num_parts"),
-            aws_json_value_new_number(meta_request->allocator, (double)auto_ranged_put->synced_data.total_num_parts));
+    if (!auto_ranged_put->synced_data.complete_multipart_upload_sent) {
```

Comment:
we can still get into weird scenarios where we've only *constructed* the final CompleteMultipartUpload request, but it hasn't actually finished yet...

we report error REQUEST_COMPLETED ... but that final request still might fail

once the meta-request is in this state, with a CompleteMultipartUpload in flight, it's basically unpausable. It must complete either successfully or unsuccessfully, but we're unable to pause it and resume later

This is a real pickle. I can't think of any synchronous solution. The only solution I can come up with is delivering the pause-token asynchronously, so that we can be 100% certain any pending requests get cancelled or finished before the pause token is delivered to the user

Reply:
if we define the pause operation to work only during the window when crt is actively uploading parts, then i dont really see the issue with throwing error when we started complete. its just how operation works by definition and its synchronous.

lets discuss more in depth. i have a couple concerns with making it async, cause i dont think its a good user experience and as mentioned above i dont really see no supporting pause when complete started as a deal breaker, cause from customer perspective there is nothing to pause anymore, since all data is already uploaded